command/server/tls_util.go

```diff
@@ -0,0 +1,167 @@
+package server
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"strings"
+	"time"
+)
+
+type CaCert struct {
+	PEM      string
+	Template *x509.Certificate
+	Signer   crypto.Signer
+}
+
+// GenerateCert creates a new leaf cert from provided CA template and signer
+func GenerateCert(caCertTemplate *x509.Certificate, caSigner crypto.Signer) (string, string, error) {
+	// Create the private key
+	signer, keyPEM, err := privateKey()
+	if err != nil {
+		return "", "", err
+	}
+
+	// The serial number for the cert
+	sn, err := serialNumber()
+	if err != nil {
+		return "", "", err
+	}
+
+	// Create the leaf cert
+	template := x509.Certificate{
+		SerialNumber:          sn,
+		Subject:               pkix.Name{CommonName: "Vault server"},
```

Comment:
I don't think this is quite right? I think we'd want to support `localhost` or `$HOSTNAME` as the common name and add the other as a DNS SANs? I think perhaps `$HOSTNAME` in CN makes most sense, and then add all the localhost variants as DNS SANs (`localhost`, `localhost4`, `localhost6`, `localhost.localdomain`, ...)? 

That way we do hostname verification correctly, independent of client. :-)

Reply:
Added!

Reply:
Thanks! I see the IPAddr SAN too (skipped over it on the first pass, lol!) so that all looks great!